lib/stdlib/src/edlin_context.erl

```diff
@@ -0,0 +1,461 @@
+        {" nuf"++Bef1, ModOrFun} -> over_fun_function(Bef1, "fun "++ModOrFun ++ Acc);
+        _ -> {Bef,Acc}
+    end.
+
+%% Extracts everything within the quote
+over_to_opening_quote(Q, Bef) when Q == $'; Q == $" ->
+    over_to_opening_quote([Q], Bef, [Q]);
+over_to_opening_quote(_, Bef) -> {Bef, []}.
+over_to_opening_quote([], Bef, Word) -> {Bef, Word};
+over_to_opening_quote([Q|Stack], [Q|Bef], Word) ->
+    over_to_opening_quote(Stack, Bef, [Q| Word]);
+over_to_opening_quote([Q|Stack], [Q,$\\|Bef], Word) ->
+    over_to_opening_quote([Stack], Bef, [$\\,Q| Word]);
+over_to_opening_quote([Stack], [C|Bef], Word) ->
+    over_to_opening_quote([Stack], Bef, [C| Word]);
+over_to_opening_quote(_,_,_) -> {no, [], []}.
+
+matching_paren($(,$)) -> true;
+matching_paren($[,$]) -> true;
+matching_paren(${,$}) -> true;
+matching_paren($<,$>) -> true;
+matching_paren(_,_) -> false.
+
+%% Extracts everything within the brackets
+%% Recursively extracts nested bracket expressions.
+over_to_opening_paren(CC, Bef) when CC == $); CC == $];
+                                    CC == $}; CC == $> ->
+    over_to_opening_paren([CC], Bef, [CC]);
+over_to_opening_paren(_, Bef) -> {Bef, []}. %% Not a closing parenthesis
+over_to_opening_paren([], Bef, Word) -> {Bef, Word};
+over_to_opening_paren(_, [], Word) -> {lists:reverse(Word), []}; %% Not a closing parenthesis
+over_to_opening_paren([CC|Stack], [CC,$$|Bef], Word) ->
+    over_to_opening_paren([CC|Stack], Bef, [$$,CC|Word]);
+over_to_opening_paren([CC|Stack], [OC|Bef], Word) when OC==$(; OC==$[; OC==${; OC==$< ->
+    case matching_paren(OC, CC) of
+        true -> over_to_opening_paren(Stack, Bef, [OC|Word]);
+        false -> over_to_opening_paren([CC|Stack], Bef, [OC|Word])
+    end;
+over_to_opening_paren([CC|Stack], [CC|Bef], Word) -> %% Nested parenthesis of same type
+    over_to_opening_paren([CC,CC|Stack], Bef, [CC|Word]);
+over_to_opening_paren(Stack, [Q|Bef], Word) when Q == $"; Q == $' ->
+    %% Consume the whole quoted text, it may contain parenthesis which
+    %% would have confused us.
+    {Bef1, QuotedWord} = over_to_opening_quote(Q, Bef),
+    over_to_opening_paren(Stack, Bef1, QuotedWord ++ Word);
+over_to_opening_paren(CC, [C|Bef], Word) -> over_to_opening_paren(CC, Bef, [C|Word]).
+
+%% Extract a whole keyword expression
+%% Keyword<code>end
+%% Function expects a string of erlang code in reverse, and extracts everything
+%% including a keyword being one of if, fun, case, maybe, receiver (need to add all here)
+%% Recursively extracts nested keyword expressions
+%% Note: In the future we could autocomplete case expressions by looking at the
+%% return type of the expression.
+over_keyword_expression(Bef) ->
+    over_keyword_expression(Bef, []).
+over_keyword_expression("dne"++Bef, Expr)->
+    %% Nested expression
+    {Bef1, KWE}=over_keyword_expression(Bef),
+    over_keyword_expression(Bef1, KWE++"end"++Expr);
+over_keyword_expression("fi"++Bef, Expr) -> {Bef, "if" ++ Expr};
```

Comment:
I checked with @bjorng and there is no such function. 

We are only interested in keywords that end with an 'end' atom, so we would have to keep a list of those as well. Maybe it would be more concise if we tried to read a word and check if it is one of our special keywords, instead of pattern matching on the reversed string.  Do you think its worth adding those two reserved words functions, and doing the refactoring?